### PR TITLE
Gc idle

### PR DESCRIPTION
**Pull request by rabbitonweb**
Thank you for contributing to ENSIME!

If this is your first time submitting a PR, this is what you can expect:

1. You'll have to agree to the licences we use. You'll get a link to follow in a moment.
2. Your code will be automatically tested. That can take as long as an hour before the results come back. You can read the `.drone.yml` and `appveyor.yml` files to see what is being run.
3. We'll try to review your code and give feedback quickly.
4. Please squash your commits into independent units. We use the commit log as our changelog so it is good if your commit messages are clear and concise.
5. You can request an ENSIME sticker from `@fommil`, be sure to ask if you want one!
6. If everything goes silent, please ping us.

Now, delete this line and everything above it, then tell us which tickets this PR closes (please don't reference tickets directly from your commit messages).

### CODE REVIEW

**Comment by rabbitonweb** [comment]
Missing unit tests

File: core/src/main/scala/org/ensime/core/IdleScheduler.scala [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+Actor that sends Idle  message every times seconds to Project actor.
+ */
+class IdleScheduler(project: ActorRef) extends Actor {
+
+  import context.{ dispatcher, system }
+
+  var idleCallback: Option[Cancellable] = None
+
+  override def preStart(): Unit = {
+    idleCallback = Some(context.system.scheduler.schedule(5.seconds, 5.seconds, project, Idle))
```
Comment (fommil) [inline core/src/main/scala/org/ensime/core/IdleScheduler.scala]: can you put this in the server config? 5 seconds is way too short as well, probably every 1 minute.

File: core/src/main/scala/org/ensime/core/Project.scala [hunk 2]
```diff
@@ -55,6 +55,20 @@ class Project(
   }
   private val classfileWatcher = context.actorOf(Props(new ClassfileWatcher(config, searchService :: reTypecheck :: Nil)), "classFileWatcher")
 
+  var seenWorkSinceIdle = true
+  //  var lastMessageReceived = System.currentTimeMillis()
```
Comment (fommil) [inline core/src/main/scala/org/ensime/core/Project.scala]: commented code, please remove

File: core/src/main/scala/org/ensime/core/Project.scala [hunk 3]
```diff
@@ -55,6 +55,20 @@ class Project(
...
+
+  def processIdleMessage(): Unit = {
+    if (seenWorkSinceIdle)
+      seenWorkSinceIdle = false
+    else
+      System.gc()
```
Comment (fommil) [inline core/src/main/scala/org/ensime/core/Project.scala]: you need to run finalizers as well, see https://github.com/fommil/sbt-big-project/blob/master/src/main/scala/BigProjectSettings.scala#L106-L108

File: core/src/main/scala/org/ensime/core/Project.scala [hunk 4]
```diff
@@ -114,7 +128,13 @@ class Project(
...
   private var rechecking: Cancellable = _
 
-  def handleRequests: Receive = withLabel("handleRequests") {
+  def idleMonitor(r: Receive)(implicit context: ActorContext): Receive = {
+    case i @ Idle =>
+      processIdleMessage(); r(i)
+    case i => seenWorkMessage; r(i)
+  }
+
+  def handleRequests: Receive = withLabel("handleRequests")(idleMonitor {
```
Comment (fommil) [inline core/src/main/scala/org/ensime/core/Project.scala]: the more canonical thing to do here would be `idleMonitor orElse handleRequests` when you set the listening state of the actor.

File: core/src/main/scala/org/ensime/core/Project.scala [hunk 5]
```diff
@@ -55,6 +55,20 @@ class Project(
   }
   private val classfileWatcher = context.actorOf(Props(new ClassfileWatcher(config, searchService :: reTypecheck :: Nil)), "classFileWatcher")
 
+  var seenWorkSinceIdle = true
```
Comment (fommil) [inline core/src/main/scala/org/ensime/core/Project.scala]: should this not be `false` by default?

**Comment by fommil** [comment]
thanks for putting a PR together! A couple of comments in the commits.

One thing that comes to mind is that we might have a cleaner way of implementing this logic in the `Project`. Instead of recording if we have seen anything since the last idle message, why not just fire off a `Cancellable` every time we get a message?

**Comment by fommil** [comment]
@rabbitonweb ! Ping

and more thoughts on this? I might be picking up something in this area on Monday if I get round to it.

**Comment by fommil** [comment]
I'm thinking that remote GC through REST is probably a better way to do this. Would you like to raise a new PR for that or are you OK if I work on it?

**Pull request closed without merging**